flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierUnaligner.java

```diff
@@ -387,12 +378,28 @@ synchronized void resetReceivedBarriers(long checkpointId) {
 			return allBarriersReceivedFuture;
 		}
 
-		synchronized void onChannelClosed() {
+		synchronized boolean onChannelClosed() throws IOException {
 			numOpenChannels--;
+
+			if (numBarriersReceived > 0) {
+				resetReceivedBarriers();
+				notifyAbort(new CheckpointException(CheckpointFailureReason.CHECKPOINT_DECLINED_INPUT_END_OF_STREAM));
+				return true;
+			}
+			return false;
 		}
 
-		synchronized void setCurrentReceivedCheckpointId(long currentReceivedCheckpointId) {
-			this.currentReceivedCheckpointId = Math.max(currentReceivedCheckpointId, this.currentReceivedCheckpointId);
+		synchronized boolean setCancelledCheckpointId(long canceledCheckpointId) {
+			boolean shouldAbort = false;
+			if (canceledCheckpointId > currentReceivedCheckpointId) {
+				currentReceivedCheckpointId = canceledCheckpointId;
+				shouldAbort = true;
+
+			} else if (canceledCheckpointId == currentReceivedCheckpointId && isCheckpointPending()) {
+				resetReceivedBarriers();
+				shouldAbort = true;
+			}
+			return shouldAbort;
```

Comment:
I think something is missing in the:
```
	private void notifyCheckpoint(CheckpointBarrier barrier) throws IOException {
		// ignore the previous triggered checkpoint by netty thread if it was already canceled or aborted before.
		if (barrier.getId() >= threadSafeUnaligner.getCurrentCheckpointId()) {
			super.notifyCheckpoint(barrier, 0);
		}
	}
```
? As it is now, after cancelling/closing checkpoint, the above method would still trigger the checkpoint, as `barrier.getId()` would be equal to `currentReceivedCheckpointId`

Reply:
For cancellation checkpoint case, it is still valid because the the `currentReceivedCheckpointId` from `threadSafeUnaligner.getCurrentCheckpointId()` would be updated via `#setCancelledCheckpointId()`.

For end of partition case, we did not couple it with specific checkpoint id as explained in above comment. And we only abort the current pending checkpoint if have, and allow the following new checkpoint happen afterwards.